File: src/main/python/Day14.py
import Utils as Utils

def processCount(line, rules, count):
    patterns = dict()
    chars = [ 0 ] * 26
    
    for i in range(len(line)):
        if i + 1 < len(line):
            key = line[i:i+2]
            patterns[key] = patterns.get(key, 0) + 1
        chars[ord(line[i]) - 65] += 1
    
    for i in range(count):
        nextPatterns = dict()
        for key, amount in patterns.items():
            ruleCharacter = rules[key]
            aa = key[0] + ruleCharacter
            bb = ruleCharacter + key[1]
            nextPatterns[aa] = nextPatterns.get(aa, 0) + amount
            nextPatterns[bb] = nextPatterns.get(bb, 0) + amount
            chars[ord(ruleCharacter) - 65] += amount
        patterns = nextPatterns
    
    return max(chars) - min(x for x in chars if x != 0)

@Utils.time_it_and_evaluate
def partOne(line, rules):
    return processCount(line, rules, 10)

@Utils.time_it_and_evaluate
def partTwo(line, rules):
    return processCount(line, rules, 40)

if __name__ == '__main__':
    lines = Utils.readAllLines(2021, 'day14')
    
    line = lines[0]
    rules = dict()
    for x in lines[2:]:
        key, val = x.split(' -> ')
        rules[key] = val
    
    print('Day 14')
    partOne(line, rules)
    partTwo(line, rules)

File: src/main/java/me/hardcoded/aoc2021/Day14.java
package me.hardcoded.aoc2021;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import me.hardcoded.util.Utils;

public class Day14 {
	public static void main(String[] args) throws Exception {
		List<String> lines = Utils.readAllLines(2021, "day14");
		
		String line = lines.get(0);
		Map<String, Character> rules = lines.stream().skip(2).map(s -> s.split(" -> ")).collect(Collectors.toMap(v -> v[0], v -> v[1].charAt(0)));
		
		Utils.printf("Day 14\n");
		Utils.printf("PartOne: %d\n", partOne(line, rules));
		Utils.printf("PartTwo: %d\n", partTwo(line, rules));
	}
	
	private static long processCount(String line, Map<String, Character> rules, int count) {
		Map<String, Long> patterns = new HashMap<>();
		Map<String, Long> nextPatterns = new HashMap<>();
		long[] chars = new long[26];
		
		for(int i = 0; i < line.length(); i++) {
			if(i + 1 < line.length()) {
				String str = line.substring(i, i + 2);
				patterns.put(str, patterns.getOrDefault(str, 0L) + 1L);
			}
			chars[line.charAt(i) - 'A']++;
		}
		
		for(int i = 0; i < count; i++) {
			nextPatterns.clear();
			
			for(Map.Entry<String, Long> entry : patterns.entrySet()) {
				String str = entry.getKey();
				
				char add = rules.get(str);
				String aa = str.charAt(0) + "" + add;
				String bb = add + "" + str.charAt(1);

				long amount = entry.getValue();
				nextPatterns.put(aa, nextPatterns.getOrDefault(aa, 0L) + amount);
				nextPatterns.put(bb, nextPatterns.getOrDefault(bb, 0L) + amount);
				chars[add - 'A'] += amount;
			}
			
			patterns.clear();
			patterns.putAll(nextPatterns);
		}
		
		long max = Long.MIN_VALUE;
		long min = Long.MAX_VALUE;
		for(int i = 0; i < 26; i++) {
			max = Math.max(max, chars[i]);
			if(chars[i] != 0) min = Math.min(min, chars[i]);
		}
		
		return max - min;
	}
	
	// Solve: 9 min
	public static long partOne(String line, Map<String, Character> rules) throws Exception {
		return processCount(line, rules, 10);
	}
	
	// Solve 25 min
	public static long partTwo(String line, Map<String, Character> rules) throws Exception {
		return processCount(line, rules, 40);
	}
}
